tests/core/snapd-failover/task.yaml

```diff
@@ -1,64 +1,163 @@
 
-    echo "And verify that snap commands still work and snapd is reverted"
-    retry-tool -n 30 --wait 5 systemctl is-active snapd.failure.service
-    snap list | MATCH "^snapd .* $current .*"
+        # test that random signals don't trigger snapd-failure
+        current=$(readlink /snap/snapd/current)
 
-    echo "Verify we got the expected error message"
-    snap change --last=install|MATCH "there was a snapd rollback across the restart"
+        echo "Verify that a random signal does not trigger the failure handling"
+        echo "and snapd is just restarted"
+        systemctl kill --signal=SIGSEGV snapd.service
 
-    echo "restart snapd and ensure we can still talk to it"
-    systemctl restart snapd.socket snapd.service
-    snap list | MATCH "^snapd .* $current .*"
+        started_after_rand_sig="$(get_journalctl_log -u snapd.failure | grep -c 'Started Failure handling of the snapd snap.' || true)"
+
+        if [ "$started_after_rand_sig" -gt "$started_before" ] ; then
+            echo "snapd.failure inadvertently started with a random signal!"
+            exit 1
+        fi
+
+        echo "Snap list is working still"
+        snap list | MATCH "^snapd .* $current .*"
+
+        SNAPD_SNAP=$(ls /var/lib/snapd/snaps/snapd_"$current".snap)
+
+        echo "Break snapd"
+        unsquashfs -d ./snapd-broken "$SNAPD_SNAP"
+        echo "" > ./snapd-broken/usr/lib/snapd/snapd
+        (cd ./snapd-broken && snap pack .)
+
+        # try to install a broken snapd snap multiple times, so that if a device
+        # wants to keep trying a refresh to a broken snapd snap for some reason
+        # it will be able to revert and fallback as many times as needed
+        #shellcheck disable=SC2167
+        for _ in $(seq 1 2); do
+            started_before="$(get_journalctl_log -u snapd.failure | grep -c 'Started Failure handling of the snapd snap.' || true)"
+
+            echo "Install the broken snapd"
+            if snap install --dangerous ./snapd-broken/snapd_*.snap; then
+                echo "installing a broken snapd should not work, test broken"
+                exit 1
+            fi
+
+            echo "Verify that snapd.failure was activated when we tried to install a broken snapd"
+            #shellcheck disable=SC2165
+            for _ in $(seq 60); do
+                # get the number of times that snapd.failure was started
+                started_after="$(get_journalctl_log -u snapd.failure | grep -c 'Started Failure handling of the snapd snap.' || true)"
+                if [ "$started_after" -gt "$started_before" ] ; then
+                    break
+                fi
+                sleep 1
+            done
+
+            # just because snapd.failure.service is active doesn't mean that we are 
+            # fully ready; we should wait until the snap command shows up again
+            echo "And verify that snap commands still work and snapd is reverted"
```

Comment:
Do you think that snapd-failure.service could be a one shot service that we just wait for here?

Reply:
1. It currently is a one-shot service but it has `RemainAfterExit=true`, so that's what we used to wait for here
2. We shouldn't have it set `RemainAfterExit=true`, because this way we run into the bug where trying to restore snapd twice during the same boot doesn't trigger `snapd.failure.service` to re-run, and we are broken.
3. We could refactor to still keep `RemainAfterExit=true`, but have snapd itself stop and set `snapd.failure.service` to be inactive when it has been restored, but this is racy and a bit difficult to reason about, so I think the current solution is better.